libyul/optimiser/UnusedFunctionParameterPruner.cpp

```diff
@@ -0,0 +1,136 @@
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * UnusedFunctionParameterPruner: Optimiser step that removes unused parameters from function
+ * definition.
+ */
+
+#include <libyul/optimiser/UnusedFunctionParameterPruner.h>
+#include <libyul/optimiser/UnusedFunctionsCommon.h>
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/optimiser/NameDisplacer.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/YulString.h>
+#include <libyul/AsmData.h>
+#include <libyul/Dialect.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <algorithm>
+#include <optional>
+#include <variant>
+
+using namespace std;
+using namespace solidity::util;
+using namespace solidity::yul;
+
+namespace
+{
+
+bool anyFalse(vector<bool> const& _mask)
+{
+	return any_of(_mask.begin(), _mask.end(), [](bool b){ return !b; });
+}
+
+template <typename K, typename V>
+optional<V> optionalFind(map<K, V> const& _map, K const& _key)
+{
+	auto it = _map.find(_key);
+
+	if (it != _map.end())
+		return it->second;
+
+	return nullopt;
+}
+
+
+} // anonymous namespace
+
+void UnusedFunctionParameterPruner::run(OptimiserStepContext& _context, Block& _block)
+{
+
+	map<YulString, size_t> references = ReferencesCounter::countReferences(_block);
+	auto used = [&](auto v) -> bool { return references.count(v.name); };
+
+	// Function name and a boolean mask, where `false` at index `i` indicates that the function
+	// argument at index `i` in `FunctionDefinition::parameters` is unused inside the function body.
+	map<YulString, vector<bool>> usedParameters;
+	// Function name and a boolean mask, where `false` at index `i` indicates that the
+	// return-parameter at index `i` in `FunctionDefinition::returnVariables` is unused inside the
+	// function body.
+	map<YulString, vector<bool>> usedReturnVariables;
+
+	// Step 1 of UnusedFunctionParameterPruner: Find functions whose parameters (both arguments and
+	// return-parameters) are not used in its body.
+	for (auto const& statement: _block.statements)
+		if (holds_alternative<FunctionDefinition>(statement))
+		{
+			FunctionDefinition const& f = std::get<FunctionDefinition>(statement);
+
+			if (wasPruned(f, _context.dialect))
+				continue;
+
+			vector<bool> parameters = applyMap(f.parameters, used);
+			vector<bool> returnVariables = applyMap(f.returnVariables, used);
+
+			if (anyFalse(parameters))
+				usedParameters[f.name] = move(parameters);
+			if (anyFalse(returnVariables))
+				usedReturnVariables[f.name] = move(returnVariables);
+		}
+
+	set<YulString> namesToFree = util::keys(usedParameters) + util::keys(usedReturnVariables);
+
+	// Step 2 of UnusedFunctionParameterPruner: Replace all references of the function, say `f`, by
+	// a new name, say `f_1`.
```

Comment:
This also renames the function itself, right?

Reply:
Yes. I'll include it in the comment.